59.md

```diff
@@ -6,157 +6,193 @@ Gift Wrap
 
-Relays may choose not to store gift wrapped events due to them not being publicly useful. Clients MAY choose
+Relays MAY choose not to store Gift-wrapped events due to them not being publicly useful. Clients MAY choose
 to attach a certain amount of proof-of-work to the wrapper event per [NIP-13](13.md) in a bid to demonstrate that
 the event is not spam or a denial-of-service attack.
 
-To protect recipient metadata, relays SHOULD guard access to `kind 1059` events based on user AUTH. When
-possible, clients should only send wrapped events to relays that offer this protection.
+# Privacy-first Protocol
+
+In cases of elevated privacy needs, the 3 primitives can be used together, one after another. In this scheme, even if all of the decrypted payloads leak, the sender and receiver are not publicly connected in a verifiable event and none of the inner event's information can be publicly verified.
+
+To transfer an event from pubkey A to pubkey B, perform the following steps:   
+1. Turn an event into a Rumor by deleting the signature. 
+2. Seal the Rumor 
+3. Gift Wrap the Seal 
+4. Send the Gift Wrap to the relays of the receiver.
+
+In pseudo-code:
+
+```js
+{
+  "id": "<id>",
+  "pubkey": "<random-pubkey>",
+  "created_at": randomTimeUpTo2DaysInThePast()
+  "kind": 1059,
+  "tags": [
+    ["p", "<pubkey-B>"],
+  ],
+  "content": nip44Encrypt({
+    "id": "<id>",
+    "pubkey" "<pubkey-A>",
+    "created_at": "created_at": randomTimeUpTo2DaysInThePast()
+    "kind": 13, // seal
+    "tags": [],
+    "content": nip44Encrypt(rumor, pubkeyB, privkeyA),
+    "sig": "<signed by pubkey-A>"
+  }, pubkeyB, privkeyRandom),
+  "sig": "<random, one-time-use pubkey signature>"
+}
+```
+
+If the `rumor` is intended for more than one party, or if the author wants to retain an encrypted copy, a single `rumor` may be sealed, wrapped and addressed for each recipient and the author individually.
+
+The canonical `created_at` time belongs to the `rumor`. All other timestamps SHOULD be tweaked to thwart time-analysis attacks. Note that some relays don't serve events dated in the future, so all timestamps SHOULD be in the past.
+
+# Gift Wrap as an onion routing payload
+
+Gift wraps can be used separetedly to hide any signed event from the public and routing relays/servers. 
+
+For instance, to send an event through a push notification stack without the stack knowing about the event or who is it for, simply wrap it on a `kind 1059` event. Since the stack already has a token, the `p` tag to the receiver (which would reveal who the receiver is) is not needed.  
+
+```js
+{
+  "id": "<id>",
+  "pubkey": "<random-pubkey>",
+  "created_at": randomTimeUpTo2DaysInThePast()
+  "kind": 1059,
+  "tags": [
+    ["token", "<destination's push notification token>"],
+  ],
+  "content": nip44Encrypt(notificationEvent, pubkeyB, privkeyRandom)
+  "sig": "<random, one-time-use pubkey signature>"
+}
+```
```

Comment:
I don't think this needs to be included.

Reply:
I don't know.. the text without it feels like people are required to do rumor -> seal -> giftwrap which is not the case at all. Otherwise, every example is an example of the rumor -> seal -> giftwrap scheme, which I am starting to think it should be only used for DMs and other super private stuff. The average case for GiftWraps is much bigger than the private protocol.

Reply:
Somebody could even use Seals with the receiver info in it for use cases where the sender and receiver can be connected but the rest of the event can never leak.